File: src/main/org/nlogo/prim/_random.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim

import org.nlogo.api.Syntax
import org.nlogo.nvm.{ Reporter, Context }

class _random extends Reporter {

  override def syntax =
    Syntax.reporterSyntax(
      Array(Syntax.NumberType),
      Syntax.NumberType)

  override def report(context: Context) =
    Double.box(report_1(context, argEvalDoubleValue(context, 0)))

  def report_1(context: Context, maxDouble: Double): Double = {
    var maxLong = validLong(maxDouble)
    if (maxDouble != maxLong)
      maxLong += (if (maxDouble >= 0) 1 else -1)
    if (maxLong > 0)
      context.job.random.nextLong(maxLong).toDouble
    else if (maxLong < 0)
      (-context.job.random.nextLong(-maxLong)).toDouble
    else 0
  }

}

File: netlogo-gui/src/main/prim/_random.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim;

import org.nlogo.api.LogoException;
import org.nlogo.core.Syntax;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.Reporter;

public final class _random extends Reporter {


  @Override
  public Object report(Context context) throws LogoException {
    return report_1(context, argEvalDoubleValue(context, 0));
  }

  public double report_1(final Context context, double maxdouble) throws LogoException {
    long maxlong = validLong(maxdouble, context);
    if (maxdouble != maxlong) {
      maxlong += (maxdouble >= 0) ? 1 : -1;
    }
    if (maxlong > 0) {
      return context.job.random.nextLong(maxlong);
    }
    if (maxlong < 0) {
      return -context.job.random.nextLong(-maxlong);
    }
    return 0;
  }
}
